lib/bigdecimal/math.rb

```diff
@@ -170,7 +330,215 @@ def atan(x, prec)
+
+    prec += BigDecimal.double_fig
+    prec2 = prec + [-x.exponent, 0].max
+    e = BigMath.exp(x, prec2)
+    einv = BigDecimal(1).div(e, prec2)
+    (e - einv).div(e + einv, prec)
+  end
+
+  # call-seq:
+  #   asinh(decimal, numeric) -> BigDecimal
+  #
+  # Computes the inverse hyperbolic sine of +decimal+ to the specified number of digits of
+  # precision, +numeric+.
+  #
+  # If +decimal+ is NaN, returns NaN.
+  #
+  #   BigMath.asinh(BigDecimal('1'), 16).to_s
+  #   #=> "0.881373587019543025232609324892919887466177636058e0"
+  #
+  def asinh(x, prec)
+    raise ArgumentError, "Zero or negative precision for tanh" if prec <= 0
+    return x if x.nan? || x.infinite?
+    return -asinh(-x, prec) if x < 0
+
+    sqrt_prec = prec + [-x.exponent, 0].max
+    BigMath.log(x + sqrt(x**2 + 1, sqrt_prec), prec)
+  end
+
+  # call-seq:
+  #   acosh(decimal, numeric) -> BigDecimal
+  #
+  # Computes the inverse hyperbolic cosine of +decimal+ to the specified number of digits of
+  # precision, +numeric+.
+  #
+  # If +decimal+ is NaN, returns NaN.
+  #
+  #   BigMath.acosh(BigDecimal('2'), 16).to_s
+  #   #=> "0.1316957896924816708625046347239934461496535769096e1"
+  #
+  def acosh(x, prec)
+    raise ArgumentError, "Zero or negative precision for tanh" if prec <= 0
+    raise Math::DomainError, "Out of domain argument for acosh" if x < 1
+    return BigDecimal::INFINITY if x.infinite?
+    return BigDecimal::NAN if x.nan?
+
+    BigMath.log(x + sqrt(x**2 - 1, prec), prec)
+  end
+
+  # call-seq:
+  #   atanh(decimal, numeric) -> BigDecimal
+  #
+  # Computes the inverse hyperbolic tangent of +decimal+ to the specified number of digits of
+  # precision, +numeric+.
+  #
+  # If +decimal+ is NaN, returns NaN.
+  #
+  #   BigMath.atanh(BigDecimal('0.5'), 16).to_s
+  #   #=> "0.54930614433405484569762261846126e0"
+  #
+  def atanh(x, prec)
+    raise ArgumentError, "Zero or negative precision for tanh" if prec <= 0
+    raise Math::DomainError, "Out of domain argument for atanh" if x < -1 || x > 1
+    return BigDecimal::NAN if x.nan?
+    return BigDecimal::INFINITY if x == 1
+    return -BigDecimal::INFINITY if x == -1
+
+    prec += BigDecimal.double_fig
+    (BigMath.log(x + 1, prec) - BigMath.log(1 - x, prec)).div(2, prec)
+  end
+
+  # call-seq:
+  #   BigMath.log2(decimal, numeric)    -> BigDecimal
+  #
+  # Computes the base 2 logarithm of +decimal+ to the specified number of
+  # digits of precision, +numeric+.
+  #
+  # If +decimal+ is zero or negative, raises Math::DomainError.
+  #
+  # If +decimal+ is positive infinity, returns Infinity.
+  #
+  # If +decimal+ is NaN, returns NaN.
+  #
+  #   BigMath.log2(BigDecimal('3'), 16).to_s
+  #   #=> "0.158496250072115618145373894394782e1"
+  #
+  def log2(x, prec)
+    raise ArgumentError, "Zero or negative precision for log2" if prec <= 0
+    return BigDecimal::NAN if x.nan?
+    return BigDecimal::INFINITY if x.infinite? == 1
+
+    prec2 = prec + BigDecimal.double_fig * 3 / 2
```

Comment:
Just curious: why do we decide what `prec2` is needed? It differs between the functions and I'm wondering how it's calculated.

Reply:
Normally, BigMath functions calculate with `prec+BigDecimal.double_fig` precision.

I think `BigMath.log2(1024, prec)` is expected to return exactly `10`. To calculate exact value for `2**n` numbers, we need to round the result to precision = `prec+BigDecimal.double_fig`.
`log(x)/log(2)` should be calculated with `prec + BigDecimal.double_fig + extra` precision.
I decided to use `extra=BigDeciaml.double_fig/2` here.